README.md
# web-security

Decoupled web security implementations for Clojure.

[![Clojars Project](https://img.shields.io/clojars/v/com.github.strojure/web-security.svg)](https://clojars.org/com.github.strojure/web-security)

[![cljdoc badge](https://cljdoc.org/badge/com.github.strojure/web-security)](https://cljdoc.org/d/com.github.strojure/web-security)
[![tests](https://github.com/strojure/web-security/actions/workflows/tests.yml/badge.svg)](https://github.com/strojure/web-security/actions/workflows/tests.yml)

## Motivation

Provide [web security] implementations decoupled from any of http abstractions
like ring, pedestal etc.

[web security]: https://developer.mozilla.org/en-US/docs/Web/Security

## API

### Content Security Policy (CSP)

Functions implementing CSP:

- [csp/header-name] returns normal or report-only name of the CSP header.
- [csp/header-value-fn] builds function for CSP header value from policy map,
  supports dynamic nonce substitution.
- [csp/requires-nonce?] tests if result of the `header-value-fn` requires nonce
  argument
- [csp/find-directive] return policy directive value from the policy map
- [csp/random-nonce-fn] returns function to generate nonce values to be used in
  HTTP response.

[csp/header-name]:
https://cljdoc.org/d/com.github.strojure/web-security/CURRENT/api/strojure.web-security.csp#header-name

[csp/header-value-fn]:
https://cljdoc.org/d/com.github.strojure/web-security/CURRENT/api/strojure.web-security.csp#header-value-fn

[csp/requires-nonce?]:
https://cljdoc.org/d/com.github.strojure/web-security/CURRENT/api/strojure.web-security.csp#requires-nonce?

[csp/find-directive]:
https://cljdoc.org/d/com.github.strojure/web-security/CURRENT/api/strojure.web-security.csp#find-directive

[csp/random-nonce-fn]:
https://cljdoc.org/d/com.github.strojure/web-security/CURRENT/api/strojure.web-security.csp#random-nonce-fn

### HSTS

Functions for the [Strict-Transport-Security] response header:

- [hsts/header-value].

[Strict-Transport-Security]:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security

[hsts/header-value]:
https://cljdoc.org/d/com.github.strojure/web-security/CURRENT/api/strojure.web-security.hsts

### Referrer-Policy

Functions for the [Referrer-Policy] response header:

- [referrer-policy/header-value].

[Referrer-Policy]:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy

[referrer-policy/header-value]:
https://cljdoc.org/d/com.github.strojure/web-security/CURRENT/api/strojure.web-security.referrer-policy

---

Authored by [Sergey Trofimov](https://github.com/serioga).

[![license](https://img.shields.io/badge/license-The%20Unlicense-informational)](UNLICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## `1.3.0-41-SNAPSHOT`

Release date `UNRELEASED`

- (feat api): add `util.base64/estimated-strlen` function

## `1.2.0-38`

Release date `2023-03-28`

- (feat api): add `util.base64/url-encode-no-padding` function
- (feat api): add `util.random/url-safe-string-fn` function
- (refactor): impl `csp/random-nonce-fn` using `url-safe-string-fn`
    + BREAKING: remove 1-arity as considered useless
    + Generated nonce is shorter (22 chars) due to changed data size from 18 to
      16 bytes.

## `1.1.0-32`

Release date `2023-03-20`

- (feat): implement `hsts` response header
- (feat): implement `referrer-policy` response header

## `1.0.0-28`

Release date `2023-03-15`

- (feat api): add `csp/header-name` function
- (feat api): add `csp/header-value-fn` function
- (feat api): add `csp/requires-nonce?` function
- (feat api): add `csp/find-directive` function
- (feat api): add `csp/random-nonce-fn` function


test/strojure/web_security/csp_test.clj
(ns strojure.web-security.csp-test
  (:require [clojure.test :as test :refer [deftest testing]]
            [strojure.web-security.csp :as csp]))

(set! *warn-on-reflection* true)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(deftest header-name-t

  (test/is (= "Content-Security-Policy"
              (csp/header-name)))

  (test/is (= "Content-Security-Policy"
              (csp/header-name false)))

  (test/is (= "Content-Security-Policy-Report-Only"
              (csp/header-name true)))

  )

(deftest header-value-fn-t

  (testing "String keys"

    (test/is "default-src *"
             ((csp/header-value-fn (sorted-map "default-src" "*"))))

    (test/is "default-src *; image-src *"
             ((csp/header-value-fn (sorted-map "default-src" "*"
                                               "image-src" "*"))))

    (test/is "default-src *; image-src *"
             ((csp/header-value-fn (sorted-map "image-src" "*"
                                               "default-src" "*"))))

    )

  (testing "Keyword keys"

    (test/is "default-src *"
             ((csp/header-value-fn (sorted-map :default-src "*"))))

    (test/is "default-src *; image-src *"
             ((csp/header-value-fn (sorted-map :default-src "*"
                                               :image-src "*"))))

    (test/is "default-src *"
             ((csp/header-value-fn (sorted-map :csp/default-src "*"))))

    (test/is "default-src *; image-src *"
             ((csp/header-value-fn (sorted-map :csp/default-src "*"
                                               :csp/image-src "*"))))

    )

  (testing "String values"

    (test/is "default-src https:; image-src 'self'"
             ((csp/header-value-fn (sorted-map "default-src" "https:"
                                               "image-src" "'self'"))))

    )

  (testing "Keyword values"

    (test/is "default-src 'none'; image-src 'self'"
             ((csp/header-value-fn (sorted-map "default-src" :none
                                               "image-src" :self))))

    )

  (testing "Collection values"

    (test/is "default-src https: 'unsafe-eval' 'unsafe-inline'; object-src 'none'"
             ((csp/header-value-fn (sorted-map "default-src" ["https:" :unsafe-eval :unsafe-inline]
                                               "object-src" [:none]))))

    )

  (testing "Policy with nonce"

    (test/is "default-src 'none'; script-src 'nonce-TEST-NONCE'"
             ((csp/header-value-fn (sorted-map "default-src" :none
                                               "script-src" :nonce))
              "TEST-NONCE"))

    (test/is "script-src 'nonce-TEST-NONCE'; style-src 'self' 'nonce-TEST-NONCE'"
             ((csp/header-value-fn (sorted-map "script-src" :nonce
                                               "style-src" [:self :nonce]))
              "TEST-NONCE"))

    )

  )

(deftest requires-nonce?-t

  (test/is (not (csp/requires-nonce? (csp/header-value-fn {}))))

  (test/is (csp/requires-nonce? (csp/header-value-fn {:script-src :nonce})))

  (test/is (csp/requires-nonce? (csp/header-value-fn {:script-src :nonce
                                                      :font-src :nonce})))

  )

(deftest find-directive-t

  (test/is (= "/csp-report"
              (csp/find-directive "report-uri" {"default-src" :none
                                                "report-uri" "/csp-report"})))

  (test/is (= "/csp-report"
              (csp/find-directive "report-uri" {:default-src :none
                                                :report-uri "/csp-report"})))

  (test/is (= "/csp-report"
              (csp/find-directive "report-uri" {:csp/default-src :none
                                                :csp/report-uri "/csp-report"})))

  (test/is (= "/csp-report"
              (csp/find-directive :report-uri {"default-src" :none
                                               "report-uri" "/csp-report"})))

  (test/is (= "/csp-report"
              (csp/find-directive :csp/report-uri {"default-src" :none
                                                   "report-uri" "/csp-report"})))

  (test/is (= nil
              (csp/find-directive "report-uri" {"default-src" :none
                                                "script-src" :nonce})))

  )

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(deftest random-nonce-fn-t

  (test/is (string? ((csp/random-nonce-fn))))

  (test/is (<= 22 (count ((csp/random-nonce-fn)))))

  )

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;


test/strojure/web_security/hsts_test.clj
(ns strojure.web-security.hsts-test
  (:require [clojure.test :as test :refer [deftest]]
            [strojure.web-security.hsts :as hsts]))

(set! *warn-on-reflection* true)

(declare thrown?)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(deftest header-name-t

  (test/is (= "Strict-Transport-Security"
              hsts/header-name))

  )

(deftest header-value-t

  (test/is (= nil
              (hsts/header-value nil)))

  (test/is (= nil
              (hsts/header-value false)))

  (test/is (= "max-age=31536000"
              (hsts/header-value true)))

  (test/is (= "max-age=31536000; includeSubDomains; preload"
              (hsts/header-value "max-age=31536000; includeSubDomains; preload")))

  )

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;


test/strojure/web_security/util/base64_test.clj
(ns strojure.web-security.util.base64-test
  (:require [clojure.test :as test :refer [deftest]]
            [strojure.web-security.util.base64 :as base64]))

(set! *warn-on-reflection* true)

(declare thrown?)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(deftest estimated-strlen-t

  (test/is (= 22 (base64/estimated-strlen 16)))

  (test/is (= 43 (base64/estimated-strlen 32)))

  )

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;


src/strojure/web_security/hsts.clj
(ns strojure.web-security.hsts
  "The HTTP [Strict-Transport-Security] response header (often abbreviated as
  HSTS) informs browsers that the site should only be accessed using HTTPS, and
  that any future attempts to access it using HTTP should automatically be
  converted to HTTPS.

  [Strict-Transport-Security]:
  https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security
  ")

(set! *warn-on-reflection* true)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(def ^:const ^String header-name
  "Returns \"Strict-Transport-Security\" string."
  "Strict-Transport-Security")

(defprotocol HstsHeader
  (header-value
    ^java.lang.String [obj]
    "Returns string value for the [Strict-Transport-Security] response header.

    - \"max-age=31536000\" for boolean `true`
    - string as is for strings

    [Strict-Transport-Security]:
    https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security
    "))

(extend-protocol HstsHeader
  nil (header-value [_] nil)
  String (header-value [s] s)
  Boolean (header-value [b] (when b "max-age=31536000")))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;


src/strojure/web_security/csp_impl.clj
(ns strojure.web-security.csp-impl
  (:require [clojure.string :as string])
  (:import (clojure.lang IPersistentCollection Keyword)))

(set! *warn-on-reflection* true)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defprotocol PolicyRender
  "A protocol for rendering various types as content policy directives and their
  values."

  (as-directive-name
    [x]
    "Returns string value for directive name.")

  (write-directive-value
    [x rf to]
    "Returns new value of `to` created using reducing function `(fn rf [to string])`.
    The `x` is responsible for the space before it."))

(extend-type String
  PolicyRender
  (as-directive-name [s] s)
  (write-directive-value [s rf to] (-> to (rf " ") (rf s))))

(def ^:private ^:const nonce-placeholder "__NONCE__")

(extend-type Keyword
  PolicyRender
  (as-directive-name [k]
    (.getName k))
  (write-directive-value [k rf to]
    (if (.equals :nonce k)
      (-> to (rf " 'nonce-") (rf nonce-placeholder) (rf "'"))
      ;; Wrap keyword with single quotes.
      (-> to (rf " '") (rf (.getName k)) (rf "'")))))

(extend-type IPersistentCollection
  PolicyRender
  (write-directive-value [coll rf to]
    (reduce (fn [to x] (write-directive-value x rf to)) to coll)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn sb-append
  "Reducing function which appends string to `StringBuilder`."
  ([] (StringBuilder.))
  ([sb] (.toString ^StringBuilder sb))
  ([sb s] (.append ^StringBuilder sb ^String s)))

(defn render-header-value
  "Returns content policy header value for the `policy` map.
  See [[csp/header-value-fn]] for the description of policy map."
  [policy]
  (-> (let [empty?! (volatile! true)]
        (fn [sb k v]
          (as-> sb to
            (if (.deref empty?!)
              (do (vreset! empty?! false)
                  to)
              (sb-append to "; "))
            (sb-append to (as-directive-name k))
            (write-directive-value v sb-append to))))
      (reduce-kv (sb-append) policy)
      (sb-append)))

(let [nonce-pattern (re-pattern nonce-placeholder)]

  (defn split-nonce
    "Splits header value on nonce. Returns vector of all parts."
    [s]
    (string/split s nonce-pattern -1)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;


src/strojure/web_security/util/base64.clj
(ns strojure.web-security.util.base64
  (:import (java.util Base64)))

(set! *warn-on-reflection* true)
(set! *unchecked-math* :warn-on-boxed)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(let [encoder (-> (Base64/getUrlEncoder) .withoutPadding)]

  (defn url-encode-no-padding
    "Returns string of byte array `data` URL encoded without padding."
    {:added "1.2"}
    [^bytes data]
    (.encodeToString encoder data)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn estimated-strlen
  "Returns min length of base64 encoded string for byte array of `size` bytes,
  without padding."
  {:tag Integer :added "1.3"}
  [size]
  (int (Math/ceil (-> ^int size (/ 3.0) (* 4.0)))))

(comment
  (estimated-strlen 16) :=> 22                    ; Execution time mean : 6.270033 ns
  )

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;


src/strojure/web_security/util/random.clj
(ns strojure.web-security.util.random
  (:require [strojure.web-security.util.base64 :as base64])
  (:import (java.security SecureRandom)))

(set! *warn-on-reflection* true)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn random-string-fn
  "Returns 0-arity function to generate random byte array of `size` bytes
  encoded with function `encode`. Uses instance of `SecureRandom` to generate
  random data."
  {:added "1.2"}
  [encode, size]
  (let [random (SecureRandom.)]
    (fn []
      (let [data (byte-array size)]
        (.nextBytes random data)
        (encode data)))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn url-safe-string-fn
  "Returns 0-arity function to generate random byte array of `size` bytes which
  is URL encoded without padding (16 bytes produce 22-char string). Uses
  [[random-string-fn]] with [[base64/url-encode-no-padding]] as `encode-fn`.

  Example:

      (def url-safe-string (url-safe-string-fn 16))

      (url-safe-string)
      :=> \"CAAuS3CWaMZqh490vfoFWA\"
      ;             Execution time mean : 1.134222 µs
      ;    Execution time std-deviation : 141.224567 ns
      ;   Execution time lower quantile : 1.060196 µs ( 2.5%)
      ;   Execution time upper quantile : 1.378707 µs (97.5%)
  "
  {:added "1.2"}
  [size]
  (random-string-fn base64/url-encode-no-padding size))

(comment
  (def url-safe-string (url-safe-string-fn 16))
  (url-safe-string) :=> "CAAuS3CWaMZqh490vfoFWA"
  (count (url-safe-string)) :=> 22
  )

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;


src/strojure/web_security/referrer_policy.clj
(ns strojure.web-security.referrer-policy
  "The [Referrer-Policy] HTTP header controls how much referrer information
  (sent with the Referer header) should be included with requests.

  [Referrer-Policy]:
  https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy
  "
  (:import (clojure.lang Keyword)))

(set! *warn-on-reflection* true)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(def ^:const ^String header-name
  "Returns \"Referrer-Policy\" string."
  "Referrer-Policy")

(defprotocol ReferrerPolicyHeader
  (header-value
    ^java.lang.String [obj]
    "Returns string value for the [Referrer-Policy] response header.

    - \"strict-origin-when-cross-origin\" for boolean `true`
    - string as is for strings
    - keyword name for keywords

    [Referrer-Policy]:
    https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy
    "))

(extend-protocol ReferrerPolicyHeader
  nil (header-value [_] nil)
  String (header-value [s] s)
  Boolean (header-value [b] (when b "strict-origin-when-cross-origin"))
  Keyword (header-value [k] (name k)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;


src/strojure/web_security/csp.clj
(ns strojure.web-security.csp
  "Content Security Policy (CSP).

  See https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP.
  "
  (:require [strojure.web-security.csp-impl :as impl]
            [strojure.web-security.util.random :as random]))

(set! *warn-on-reflection* true)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
;; ## CSP Header ##

(defn header-name
  "Returns CSP header name, normal one or report-only if `report-only` is true.
  Without argument returns normal header name."
  ([] (header-name false))
  ([report-only]
   (if report-only "Content-Security-Policy-Report-Only"
                   "Content-Security-Policy")))

(defn header-value-fn
  "The returned function takes a `policy` map and returns a content policy
  header value. If the policy map contains `:nonce` in directive values, the
  returned function is a 1-arity function `(fn [nonce] policy-header)`,
  otherwise it is a 0-arity function.

  The keys in the policy map represent directive names, and can be:

  - string, which is taken as is
  - keyword, which are converted to string using `name`
  - other type which implements `as-directive-name` of [[csp-impl/PolicyRender]]
    protocol

  The values in the policy map represent directive values, and can be:

  - string, which is taken as is
  - keyword
      - the special `:nonce` keyword is a placeholder for the value of nonce
      - other keywords are converted to single quoted string using `name`
  - collection, where elements are rendered as directive values and
    separated with space.
  - other type which implements `write-directive-value` of
    [[csp-impl/PolicyRender]] protocol

  Example:

      (def nonce-policy
        (csp/header-value-fn {\"script-src\" :nonce
                              \"img-src\" \"*\"}))

      (nonce-policy \"MY-NONCE\")
      :=> \"script-src 'nonce-MY-NONCE'; img-src *\"

      (def static-policy
        (csp/header-value-fn {:default-src [\"https:\"
                                            :unsafe-eval
                                            :unsafe-inline]
                              :object-src :none})

      (static-policy)
      :=> \"default-src https: 'unsafe-eval' 'unsafe-inline'; object-src 'none'\"

  CSP header values can be tested online:

  - [CSP Evaluator](https://csp-evaluator.withgoogle.com/).
  - [The Mozilla Observatory](https://observatory.mozilla.org/).
  "
  [policy]
  (let [[s ss & more :as sss] (-> (impl/render-header-value policy)
                                  (impl/split-nonce))]
    (cond more, ^::nonce (fn multiple-nonce
                           [nonce] (transduce (interpose nonce) impl/sb-append sss))
          ss,,, ^::nonce (fn singe-nonce
                           [nonce] (-> ^String s
                                       (.concat nonce)
                                       (.concat ss)))
          :else,,,,,,,,, (fn static-string
                           [] s))))

(defn requires-nonce?
  "True if result of the [[header-value-fn]] requires nonce argument."
  [f]
  (some-> (meta f) ::nonce))

(defn find-directive
  "Returns directive value from the `policy` map for the given `policy-name`."
  [policy-name policy]
  (let [policy-name (impl/as-directive-name policy-name)
        matched-value (fn matched-value [[k v]]
                        (when (= policy-name (impl/as-directive-name k)) v))]
    (some matched-value policy)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
;; ## CSP Nonce ##

(defn random-nonce-fn
  "Returns unique random 128 bit URL-safe string (22 chars) to be used as CSP
  nonce in HTTP response. Implemented with [[util.random/url-safe-string-fn]].

  Example:

      (def random-nonce (csp/random-nonce-fn))

      (random-nonce) :=> \"AsiTZwAOG_orOX0-4Vw_7g\"
      ;             Execution time mean : 1.074505 µs
      ;    Execution time std-deviation : 63.286135 ns
      ;   Execution time lower quantile : 1.000594 µs ( 2.5%)
      ;   Execution time upper quantile : 1.162035 µs (97.5%)

  See also [Using a nonce with CSP](https://content-security-policy.com/nonce/).
  "
  []
  (random/url-safe-string-fn 16))

(comment
  (def random-nonce (random-nonce-fn))
  (random-nonce) :=> "AsiTZwAOG_orOX0-4Vw_7g"
  )

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
